File: packages/api/go/internal/shared/watchdog_test.go
package shared

import (
	"context"
	"sync/atomic"
	"testing"
	"time"
)

func TestIsOrphaned(t *testing.T) {
	cases := []struct {
		name       string
		supervisor string
		parent     int
		want       bool
	}{
		// The shell that spawned us is still our parent — the normal state, all day.
		{"parent unchanged", "4242", 4242, false},
		// The incident, verbatim: the shell was SIGKILLed and macOS handed us to launchd, where we
		// went on holding :3032 for a window that no longer had a backend.
		{"reparented to launchd", "4242", 1, true},
		// Stronger than `ppid == 1`: a subreaper adopts the orphan instead of init and this still sees it.
		{"reparented to a subreaper", "4242", 9999, true},
		// Everything below is `bun dev` / `go test` / the e2e harness: nothing is supervising us, and
		// a false positive would shut the gateway down one second into every local session.
		{"no supervisor declared", "", 1, false},
		{"blank supervisor", "   ", 1, false},
		{"malformed supervisor", "nope", 1, false},
		{"zero supervisor", "0", 1, false},
		{"negative supervisor", "-1", 1, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := IsOrphaned(tc.supervisor, tc.parent); got != tc.want {
				t.Fatalf("IsOrphaned(%q, %d) = %v, want %v", tc.supervisor, tc.parent, got, tc.want)
			}
		})
	}
}

// The loop must sit still while the parent is alive and react the moment it is not.
func TestWatchParentReactsOnlyAfterTheParentChanges(t *testing.T) {
	var parent atomic.Int64
	parent.Store(4242)
	var shutdowns atomic.Int64

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	go watchParent(ctx, "4242", 2*time.Millisecond, func() int { return int(parent.Load()) }, func() {
		shutdowns.Add(1)
	})

	time.Sleep(30 * time.Millisecond)
	if got := shutdowns.Load(); got != 0 {
		t.Fatalf("com o pai vivo o watchdog nao pode desligar nada, mas desligou %d vez(es)", got)
	}

	parent.Store(1)
	deadline := time.Now().Add(time.Second)
	for shutdowns.Load() == 0 && time.Now().Before(deadline) {
		time.Sleep(2 * time.Millisecond)
	}
	if got := shutdowns.Load(); got != 1 {
		t.Fatalf("orfao tem de pedir shutdown exatamente uma vez, pediu %d", got)
	}

	// And exactly once: the loop returns after firing, so a slow fx unwind is never raced by a
	// second Shutdown from the same watchdog.
	time.Sleep(30 * time.Millisecond)
	if got := shutdowns.Load(); got != 1 {
		t.Fatalf("o watchdog tem de parar depois de disparar, mas disparou %d vezes", got)
	}
}

// A cancelled context stops it — that is what keeps the goroutine from outliving a NORMAL shutdown
// and firing a redundant Shutdown into an app that is already unwinding.
func TestWatchParentStopsOnContextCancel(t *testing.T) {
	var shutdowns atomic.Int64
	ctx, cancel := context.WithCancel(context.Background())
	cancel()

	done := make(chan struct{})
	go func() {
		watchParent(ctx, "4242", time.Millisecond, func() int { return 1 }, func() { shutdowns.Add(1) })
		close(done)
	}()

	select {
	case <-done:
	case <-time.After(time.Second):
		t.Fatal("watchParent nao respeitou o cancelamento do contexto")
	}
	if got := shutdowns.Load(); got != 0 {
		t.Fatalf("cancelado antes do primeiro tick, nao pode ter desligado nada (%d)", got)
	}
}

File: packages/api/typescript/core/src/utils/Watchdog.test.ts
import { describe, expect, test } from 'bun:test'

import { declaredSupervisorPid, isOrphaned, startParentWatchdog } from './Watchdog'

const ALIVE = { parentPid: 4242, supervisorAlive: true }

describe('parent watchdog — the only defense that survives a SIGKILLed shell', () => {
	test('a daemon whose parent is still the shell that spawned it is not orphaned', () => {
		expect(isOrphaned('4242', ALIVE)).toBe(false)
	})

	test('a daemon reparented to launchd IS orphaned', () => {
		// This is the incident, verbatim: the shell was SIGKILLed, macOS handed the child to pid 1,
		// and it went on holding :3030 serving a stale catalog to a window that no longer had a backend.
		expect(isOrphaned('4242', { parentPid: 1, supervisorAlive: false })).toBe(true)
	})

	test('reparenting to anything else is orphaned too — the check is not `ppid === 1`', () => {
		// A subreaper (or a second shell) adopts the orphan instead of init. `ppid === 1` would miss it.
		expect(isOrphaned('4242', { parentPid: 9999, supervisorAlive: false })).toBe(true)
	})

	test('o ppid mudou mas a sonda ainda diz vivo (pid reutilizado, zumbi) — órfão mesmo assim', () => {
		expect(isOrphaned('4242', { parentPid: 1, supervisorAlive: true })).toBe(true)
	})

	test('no Windows o ppid está CONGELADO no spawn — o supervisor morto só aparece pela sonda de vida', () => {
		// Não há reparenting no Windows: `process.ppid` devolve o pid de quem criou o processo para
		// sempre, vivo ou não. Sem esta linha o watchdog nunca dispararia lá.
		expect(isOrphaned('4242', { parentPid: 4242, supervisorAlive: false })).toBe(true)
	})

	test.each([
		['unset', undefined],
		['empty', ''],
		['blank', '   '],
		['not a number', 'nope'],
		['zero', '0'],
		['negative', '-1'],
		['fractional', '12.5'],
	])('an unsupervised daemon (%s pid) is NEVER orphaned', (_label, raw) => {
		// `bun dev`, `bun test` and the e2e harness all run with no shell above them. A false positive
		// here would shut the daemon down one second into every local session.
		expect(declaredSupervisorPid(raw as string | undefined)).toBeNull()
		expect(isOrphaned(raw as string | undefined, { parentPid: 1, supervisorAlive: false })).toBe(false)
	})

	test('no supervisor declared ⇒ no watchdog at all (not even a live timer)', () => {
		expect(startParentWatchdog({ supervisorPid: undefined, intervalMs: 1, onOrphaned: () => {} })).toBeNull()
	})

	test('the reaction fires once, and only after the parent actually changes', async () => {
		let parent = 4242
		let shutdowns = 0
		const stop = startParentWatchdog({
			supervisorPid: '4242',
			currentParentPid: () => parent,
			supervisorAlive: () => true,
			onOrphaned: () => {
				shutdowns += 1
			},
			intervalMs: 2,
		})
		expect(stop).not.toBeNull()

		await Bun.sleep(20)
		expect(shutdowns, 'enquanto o pai esta vivo o watchdog nao pode fazer nada').toBe(0)

		parent = 1
		await Bun.sleep(30)
		expect(shutdowns, 'o watchdog tem de reagir dentro de poucos ticks').toBe(1)

		await Bun.sleep(20)
		expect(shutdowns, 'e uma vez so — um shutdown reentrante atropela o proprio drain').toBe(1)

		stop?.()
	})

	test('com o ppid congelado (Windows), a morte do supervisor vista pela sonda também dispara — uma vez', async () => {
		let alive = true
		let shutdowns = 0
		const stop = startParentWatchdog({
			supervisorPid: '4242',
			currentParentPid: () => 4242,
			supervisorAlive: () => alive,
			onOrphaned: () => {
				shutdowns += 1
			},
			intervalMs: 2,
		})

		await Bun.sleep(20)
		expect(shutdowns).toBe(0)

		alive = false
		await Bun.sleep(30)
		expect(shutdowns).toBe(1)

		await Bun.sleep(20)
		expect(shutdowns).toBe(1)

		stop?.()
	})

	test('a sonda default é real: um supervisor cujo pid já saiu dispara sem nenhum mock', async () => {
		// Flake teórico: reuso do pid do ghost entre `exited` e o tick do watchdog leria 'vivo'.
		// Pids são monotônicos nos SOs de CI (macOS/Linux); se este teste piscar um dia, é isso.
		const ghost = Bun.spawn([process.execPath, '-e', 'process.exit(0)'])
		await ghost.exited

		let shutdowns = 0
		const stop = startParentWatchdog({
			supervisorPid: String(ghost.pid),
			// ppid "igual" ao supervisor, como no Windows — só a sonda pode ver que ele morreu.
			currentParentPid: () => ghost.pid,
			onOrphaned: () => {
				shutdowns += 1
			},
			intervalMs: 2,
		})
		await Bun.sleep(30)
		expect(shutdowns).toBe(1)
		stop?.()
	})

	test('a sonda default é real: o nosso próprio pai (o test runner) está vivo — nada dispara', async () => {
		let shutdowns = 0
		const stop = startParentWatchdog({
			supervisorPid: String(process.ppid),
			onOrphaned: () => {
				shutdowns += 1
			},
			intervalMs: 2,
		})
		await Bun.sleep(20)
		expect(shutdowns).toBe(0)
		stop?.()
	})

	test('stopping it is enough to silence it', async () => {
		let shutdowns = 0
		const stop = startParentWatchdog({
			supervisorPid: '4242',
			currentParentPid: () => 1,
			supervisorAlive: () => false,
			onOrphaned: () => {
				shutdowns += 1
			},
			intervalMs: 2,
		})
		stop?.()
		await Bun.sleep(20)
		expect(shutdowns).toBe(0)
	})
})
